flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java

```diff
@@ -306,24 +348,17 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
         CompositeKeySerializationUtils.serializeKeyGroup(
                 keyGroupRange.getEndKeyGroup() + 1, stopKeyGroupPrefixBytes);
 
-        for (KeyedStateHandle rawStateHandle : restoreStateHandles) {
+        // Insert all remaining state through creating temporary RocksDB instances
+        for (int stateIdx = 1; stateIdx < allDownloadSpecs.size(); ++stateIdx) {
 
-            if (!(rawStateHandle instanceof IncrementalRemoteKeyedStateHandle)) {
-                throw unexpectedStateHandleException(
-                        IncrementalRemoteKeyedStateHandle.class, rawStateHandle.getClass());
-            }
+            final StateHandleDownloadSpec downloadRequest = allDownloadSpecs.get(stateIdx);
```

Comment:
This can be reverted back to foreach over `allDownloadSpecs.values()` once `allDownloadSpecs` is a map.